src/placement/ring_map.c

```diff
@@ -1385,8 +1386,49 @@ ring_obj_find_rebuild(struct pl_map *map, struct daos_obj_md *md,
 			 * for rebuild, perhaps they should be unified.
 			 */
 			if (l_shard->po_shard != -1) {
+				struct pool_target	*target;
+				int			 leader;
+
 				D_ASSERT(f_shard->rfs_tgt_id != -1);
 				D_ASSERT(idx < array_size);
+
+				/* If the caller does not care about DTX related
+				 * things (myrank == -1), then fill it directly.
+				 */
+				if (myrank == -1)
+					goto fill;
+
+				leader = pl_select_leader(md->omd_id,
+					l_shard->po_shard, layout->ol_nr,
+					true, pl_obj_get_shard, layout);
+				if (leader < 0) {
+					D_WARN("Not sure whether current shard "
+					       "is leader or not for obj "DF_OID
+					       ", fseq:%d, status:%d, ver:%d, "
+					       "shard:%d, rc:%d\n",
+					       DP_OID(md->omd_id),
+					       f_shard->rfs_fseq,
+					       f_shard->rfs_status, rebuild_ver,
+					       l_shard->po_shard, rc);
+					goto fill;
+				}
+
+				rc = pool_map_find_target(map->pl_poolmap,
+							  leader, &target);
+				D_ASSERT(rc == 1);
+
+				if (myrank != target->ta_comp.co_rank) {
+					/* The leader shard is not on current
+					 * server, then current server cannot
+					 * know whether DTXs for current shard
+					 * have been re-synced or not. So skip
+					 * the shard that will be handled by
+					 * the leader on another server.
+					 */
+					continue;
+				}
```

Comment:
I wonder if you can check the leader even before the layout is generated? since it will save a lot of time.  Anyway, we will revisit here soon for new mapless placement algothrim

Reply:
> this is only needed when re-open an "dirty container" ? So it is probably unfair to create such ULT for all open, i.e. you probably need check if the open really need dtx resync? Since we are supported yet, I agree with niu, probably just remove it for now?


But how can we know whether an container is dirty or not? Honestly, it is not easy to know whether the container need dtx resync or not until we scan the DTX table. But such scanning is just part of the DTX resync logic. In current DTX model, we will not make DTX resync when close the container, instead, we can do that when open the container, that much simplifies the DTX logic. We have re-open container test is our current daos_test cases, so we need DTX resync when open the container even if we do not support server re-integration yet.


> I wonder if you can check the leader even before the layout is generated? since it will save a lot of time. Anyway, we will revisit here soon for new mapless placement algothrim

I do not think so unless we use other leader election algorithm. The current election algorithm depends on the layout, and will elect the one with lowest failure sequence as the leader.

Reply:
I mean "dirty container" will only exist if the container is closed unexpected, then it is re-opened during reintegration, we do not support it yet anyway.  So it also means dtx has to be cleaned during container close?

Hmm, probably it would be easier to check the leader without generating the whole layout of the object once we have those new mapless layout algorithm.

Reply:
> So it also means dtx has to be cleaned during container close?

I am not sure for this, that may depends on how to define the replicas consistency. If we only guarantee that the application can get consistent data with DAOS online mode, then dtx_resync() when open the container is enough.

> Hmm, probably it would be easier to check the leader without generating the whole layout of the object once we have those new mapless layout algorithm.

mapless leader election will be much helpful for DTX model. But we still needs to avoid frequent leader switch as much as possible.